apps/extension/src/core/domains/balances/handler.ts

```diff
@@ -89,107 +89,101 @@ const subscribeBalancesByParams = async (
-  })
+  }
+  let balancesSubscriptionsGeneration = 0
+  let balancesUnsubCallbacks: Promise<UnsubscribeFn>[] = []
+
+  // create a subscription to the balances params
+  // allows us to restart the inner balances subscriptions whenever the data
+  // which they depend on changes
+  const byParamsSubscription = combineLatest([
+    // chains
+    chaindataProvider.chainsListObservable,
+    // evmNetworks
+    chaindataProvider.evmNetworksListObservable,
+    // tokens
+    chaindataProvider.tokensListObservable,
+    // miniMetadatas - not used here but we must retrigger the subscriptions when this changes
+    liveQuery(async () => await balancesDb.miniMetadatas.toArray()),
+
+    // active state of substrate chains
+    activeChainsStore.observable,
+    // active state of evm networks
+    activeEvmNetworksStore.observable,
+    // enable state of tokens
+    activeTokensStore.observable,
+  ]).subscribe({
+    next: (args) => {
+      const newSubscriptionParams = getSubscriptionParams(
+        addressesByChain,
+        addressesAndEvmNetworks,
+        addressesAndTokens,
+        ...args
+      )
+
+      // restart subscription only if params change
+      if (isEqual(subscriptionParams, newSubscriptionParams)) return
+      subscriptionParams = newSubscriptionParams
+
+      // close previous subscriptions
+      balancesSubscriptionsGeneration =
+        (balancesSubscriptionsGeneration + 1) % Number.MAX_SAFE_INTEGER
+      balancesUnsubCallbacks.forEach((cb) => cb.then((close) => close()))
 
-  let closeSubscriptionCallbacks: Promise<UnsubscribeFn>[] = []
-
-  // wait for chaindata to be hydrated,
-  // then subscribe to chaindata changes and restart subscription if params change
-  updateAndWaitForUpdatedChaindata().then(() => {
-    combineLatest([
-      // chains
-      chaindataProvider.chainsListObservable,
-      // evmNetworks
-      chaindataProvider.evmNetworksListObservable,
-      // tokens
-      chaindataProvider.tokensListObservable,
-      // miniMetadatas - not used here but we must retrigger the subscription when this changes
-      liveQuery(async () => await balancesDb.miniMetadatas.toArray()),
-      // active state of evm networks
-      activeEvmNetworksStore.observable,
-      // active state of substrate chains
-      activeChainsStore.observable,
-      // enable state of tokens
-      activeTokensStore.observable,
-    ]).subscribe({
-      next: ([
-        chains,
-        evmNetworks,
-        tokens,
-        miniMetadatas,
-        activeEvmNetworks,
-        activeChains,
-        activeTokens,
-      ]) => {
-        const newSubscriptionParams = getSubscriptionParams(
-          addressesByChain,
-          addressesAndEvmNetworks,
-          addressesAndTokens,
-          chains,
-          evmNetworks,
-          tokens,
-          activeChains,
-          activeEvmNetworks,
-          activeTokens,
-          miniMetadatas
+      const thisGeneration = balancesSubscriptionsGeneration
```

Comment:
Prevent async updates from closing-but-not-yet-closed subscriptions from being transmitted through to the frontend